File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:

        out = root.val

        def DFS(root):
            nonlocal out

            if not root: return 0

            left = DFS(root.left)
            right = DFS(root.right)
            left = max(left, 0)
            right = max(right, 0)

            out = max(out, root.val, root.val + left + right)

            return root.val + max(left, right)
        
        DFS(root)
        return out

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    int maxPathSum(TreeNode* root) {
        int out = root->val;
        DFS(root, out);
        return out;

    }

    int DFS(TreeNode* root, int& out)
    {
        if (!root) return 0;

        int left = max(DFS(root->left, out), 0);
        int right = max(DFS(root->right, out), 0);

        out = max(out, root->val + left + right);

        return root->val + max(left, right);

    }
};
